Configure requirements for a package

scheibelp opened the pull request:
Spack doesn't have an easy way to say something like "If I build package X, then I ~want~ *need* version Y":

* If you specify something on the command line, then you ensure that the constraints are applied, but the package is always built
  * Likewise if you `spack add X...` to your environment, the constraints are guaranteed to hold, but the environment always builds the package
* Conversely, you can add preferences to `packages.yaml`, but these are not guaranteed (Spack can choose other settings)

For example, you might want to specify that when you build `openmpi`, that you build `@4.0.1`: you want to allow building with other specs, but always use `@4.0.1` if building `openmpi`.

This PR adds ASP rules (and logic to `asp.py`) which can generate conditional constraints. ~My current guess is that it should be "consolidated" with `asp_spec_clauses` (or rather, that function should be made more generic to support this case).~ (EDIT: this last part is now addressed)

(Update from Jan) The schema for expressing this is a new `packages.yaml` section:

```
packages:
  zlib:
    require: '@1.2.12'                             # Only zlib 1.2.12 can be used
  libpng:
    require:
    - one_of: ['%gcc@8.4.0', '%clang@9.0.0']       # libpng can only build with one of these two compilers
    - one_of: ["@versionx", "@versiony"]
```

For any given `<package>` subsection, you can now add a `require` subsection, where you can add demands for the spec. The details you add there *must* be respected by Spack (i.e. it will either match the spec there, or fail to concretize).

TODOs:

- [x] This will require updating the YAML schema in `packages.yaml` (or some config file). (EDIT 1/7/22: that is now done) ~For now, since I'm experimenting with the ASP rules, I am manually generating what would be parsed from the config~
- [x] Tests

scheibelp committed:
Allow users to require certain constraints when building specific packages

scheibelp committed:
explanatory comments

scheibelp committed:
add example of combinatoric cfg_required constraints

scheibelp committed:
fix spec inspections for combinatoric example

scheibelp committed:
add TODO related to refactoring this to combine with asp._spec_clauses

scheibelp committed:
add comments about generic format of spec constraints

scheibelp committed:
add TODOs for improving modelling

scheibelp committed:
add note on weighting conditions; remove TODO about allowing >=1 condition group members (this already appears to be the case)

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -382,6 +382,58 @@ external_conditions_hold(Package, LocalIndex) :-
+
+%-----------------------------------------------------------------------------
+% Generalization of prior section
+%-----------------------------------------------------------------------------
```
`asp._spec_clauses` could be refactored to generate condition_groups or directly generate conditions (e.g. it would do this to apply spec constraints supplied from the command line).

scheibelp replied:
It turns out that a couple functions in `asp` (namely `condition` and `impose`) already use `_spec_clauses` to conditionally apply Spec constraints, so I can update this (and `asp.py`) to follow their lead.

scheibelp committed:
use existing framework for generating constraints from arbitrary specs (so I don't need to redefine individual constraints for version, compiler, etc.)

scheibelp committed:
force exactly-one instead of >= 1 for constraint (note that it can still be that multiple specs in a group can still match since the spec matching doesn't imply cfg_required (only the other way around)

scheibelp committed:
renaming (no logic change): removing the term 'stanza'

scheibelp committed:
edit comment

scheibelp committed:
remove unnecessary 'intermediate' variable (simplifies concretize.lp)

scheibelp commented on lib/spack/spack/solver/asp.py:
```diff
@@ -774,6 +774,38 @@ def package_compiler_defaults(self, pkg):
+            test_rules = [
+                ("openmpi", ["%gcc", "%clang"]),
+                ("openmpi", ["@4.0.1", "@4.0.0"])
+            ]
```
Proposed schema for `packages.yaml`:

```
packages:
  openmpi:
    spec_requirements:
    - require_one_of:
      - @4.0.2+static
      - @4.0.1~static
```

and

```
packages:
  openmpi:
    spec_requirements:
    - require_one_of:
      - %gcc
      - %clang
    - require_one_of:
      - @4.0.1
      - @4.0.2
    - require: +static # shorthand when you always want to enforce a given spec
```

(these correspond to the examples currently in asp.py)

tgamblin replied:
How about something like:

```yaml
packages:
  openmpi:
    require:
    - one_of: [%gcc, %clang]   # or these can be vertical
    - one_of: [@4.0.1, @4.0.2]
```

The nice thing about this is that we could add `prefer` as a sort-of equivalent to what we have now, and allow the user to set the default (require or prefer) in `all`.  I also think this is more concise (`spec_requirements` is redundant with `require_` above).

Otherwise this looks good -- the semantics seem clear to me.

scheibelp committed:
style edit

scheibelp commented:
At this point `concretize.lp` is sufficiently simple, and https://github.com/spack/spack/pull/27987#discussion_r772711706 proposes a schema structure. So @tgamblin @becker33 if that seems agreeable I can update the schema validation and tests.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -774,6 +774,34 @@ def package_compiler_defaults(self, pkg):
+        for pkg_name, stanza in test_rules:
+            self.gen.fact(fn.condition_group(pkg_name, stanza_id))
+            for spec_str in stanza:
+                spec, = spack.cmd.parse_specs(spec_str)
```
Why not just construct a `Spec` out of the `spec_str`?  That would actually force there to be one spec, while this discards any extra specs on the same line.

tgamblin commented on lib/spack/spack/solver/asp.py:
```diff
@@ -774,6 +774,34 @@ def package_compiler_defaults(self, pkg):
+                self.gen.fact(fn.condition_group_member(member_id, stanza_id))
+                self.impose(member_id, spec, name=pkg_name, node=False)
+                member_id += 1
+            stanza_id += 1
```
Nice! It looks like this is on the right track

I have not had a chance to think about the structure of this and the ASP in detail, but it'll need a rework for your TODO below anyway.  So I suggest working with @becker33 on that part.

I don't think this needs to be integrated with spec_clauses, but I could be wrong -- I think the way you've done it here, with `impose()` is good.  I don't fully understand the need for `stanza_id` *and* `member_id` -- but you will need a condition id for the `impose()`.  I think the condition can just have a single requirement -- `node(pkg_name)` and you can simplify what you have here.  @becker33 and @alalazo should be able to help with that.

scheibelp replied:
The member_id is the condition ID for `impose`, the `stanza_id` is the identifier for the group of conditions. I think it is minimal (without an identifier for the group, how can I say "one of the constraints in this group must apply"?).

scheibelp committed:
construct Spec directly vs. via parse_specs

scheibelp committed:
rename 'stanza_id' to 'cnd_grp_id' (condition group ID)

scheibelp committed:
Update packages.yaml schema and translate hard-coded examples to examples you can test with 'spack -C'

scheibelp committed:
Earlier members of condition groups are preferred

scheibelp committed:
remove root specifier from condition_group weight minimization; add todo to ask about build_priority

scheibelp commented:
This now

* Implements the schema discussed in https://github.com/spack/spack/pull/27987#discussion_r779783726 (currently there are examples placed in a `test-cfgs/` directory which will be updated to unit tests
* Implements preferences for condition groups (for a set of specs in a `one_of` specification, the earlier ones are preferred)
  * I still don't fully understand some of the logic around minimization (e.g. why do most of the minimization criteria have a `build_priority` - what is that?)

scheibelp committed:
style edit

scheibelp committed:
more style edits

scheibelp commented:
@becker33 this appears to be passing unit tests and I have mostly addressed https://github.com/spack/spack/pull/27987#pullrequestreview-845203141 minus a couple issues I have questions about: 

* I added preference weighting for conditions in condition groups, but wasn't sure about some of the mechanisms used to control/influence that like `build_priority`
* There was a request that `member_id` and `stanza_id` (now `condition_group_id`) be consolidated in https://github.com/spack/spack/pull/27987#discussion_r779907383 but I didn't think that was possible (which I explained in my response to that comment)

This still needs tests, but I wanted to make sure we agreed on the schema (now added with examples) and overall implementation first.

tgamblin commented:
@scheibelp:

> I still don't fully understand some of the logic around minimization (e.g. why do most of the minimization criteria have a build_priority - what is that?)

See the explanation here: https://github.com/spack/spack/pull/25310/commits/abda8a496f7dc203501076da76e0bbc0207135a6

If reuse is enabled, build_priority is set to some number greater than the number of optimization criteria, and it gets added to the priority of each criterion *for packages that will be built*. So if a package is to be built, we prioritize its defaults higher than minimizing the total number of builds. If a package is being reused and it satisfies any other imposed constraints, we prefer to take it as-is than to force its variants, versions, etc. to be the preferred/default ones.

For some intuition: if you don’t do this, and instead make minimizing total builds the absolute top priority, then if you have to build, say, CMake, then the solver will disable variants on the new CMake to reduce the total dependencies. E.g. it’ll give you a CMake without openssl and with other optional but still default features disabled. This is different than what a typical `spack install cmake` would do. build_priority ensures that new builds act more like typical spack installs when reuse is enabled.

scheibelp committed:
make some 'room' for inserting new constraint priorities

scheibelp committed:
relocate config requirements again to take precedence over version weights (otherwise the default version weight precedence takes priority and users can't prefer earlier versions via config requirements)

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -855,156 +870,168 @@ opt_criterion(100, "number of packages to build (vs. reuse)").
+    Weight@73+Priority
+    : condition_weight(Package, Weight),
+      build_priority(Package, Priority)
+}.
```
@becker33 we spoke about giving this the lowest priority, but if this isn't prioritized above e.g. version weights, then for example if someone has

```
require:
  one_of: [1.0, 1.1]
```

Then Spack will pick `1.1` - the user can't control this preference via `requires`. I think they could combine this with `version: [1.0, 1.1]` but IMO they should be able to control it with just the `requires` section.

That being said, if the ramifications of placing this last are understood and still desired, then I can definitely do that.

I also added some "space" between the priorities, to make it easier to insert new ones in the future.

scheibelp commented on lib/spack/spack/schema/packages.py:
```diff
@@ -69,6 +69,20 @@ def deprecate_paths_and_modules(instance, deprecated_properties):
+                             'items': {
+                                 'type': 'object',
+                                 'properties': {
+                                     'one_of': {'type': 'array'}
```
Note for reviewer: YAML schema has a concept of "oneOf" (to specify that a YAML element can have multiple types). This particular schema includes a notion of an element which is itself called "one_of", which appears in the literal YAML as:

```
packages:
  openmpi:
    require:
    - one_of: [%gcc, %clang]   # or these can be vertical
    - one_of: [@4.0.1, @4.0.2]
```

In this schema, you can either have this or something like

```
packages:
  openmpi:
    require: "%gcc" # Note this is not expressing the same thing as the previous example
```

So "require" can either be a list of "one_of" elements, or a string.

alalazo replied:
Am I correct that this schema still needs to be specified, at least for the `array` part? Also, I tried:
```yaml
packages:
  zlib:
    require: '@1.2.12'
```
and got the following parsing error:
```console
$ spack  spec libpng
==> Error: Invalid version specifier  Encountered when parsing spec:
    @
     ^
```

scheibelp replied:
This was a bug in the asp.py code that consumes the config: it is fixed by https://github.com/spack/spack/pull/27987/commits/39168f4e5f6f5c2cb6575c7af10c67e23ffdcfeb

scheibelp committed:
merge from develop (conflict in asp.py)

scheibelp committed:
merge commit conceptual drift resulted in condition groups being ignored

scheibelp commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -472,6 +472,21 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
+% at least one must be applied. Each spec in the group is associated
+% with a condition via asp.impose(), and we require that exactly one condition
+% from each condition_group holds.
+1 { condition_holds(Y) : condition_group_member(Y, X) } 1 :- node(Package), condition_group(Package, X).
```
I thought replacing this with

```prolog
condition_group_satisfied(X) :- 
    node(Package), 
    condition_group(Package, X),
    condition_group_member(Y, X),
    condition_holds(Y).

error(2, "Cannot satisfy condition group for package '{0}'", Package) :-
    node(Package),
    condition_group(Package, X),
    not condition_group_satisfied(X).
```

Would fit better (and I could also pass along more values for better error messages) but if I do

```
$ spack -C test-cfgs/example1/ spec openmpi
==> Error: Cannot satisfy condition group for package 'openmpi'
```

(i.e. now it fails: with the current `concretize.lp` logic it succeeds).

scheibelp replied:
This looks analogous to `node_os`: instead of removing

```
1 { condition_holds(Y) : condition_group_member(Y, X) } 1 :- node(Package), condition_group(Package, X).
```

I have relaxed it to

```
{ condition_holds(Y) : condition_group_member(Y, X) } :- node(Package), condition_group(Package, X).
```

I'm not sure why the `error :- ...` is needed though for cases where no condition group members apply (and likewise for OS choice).

scheibelp replied:
I think understand this better after reading [here](https://www.cs.cmu.edu/~cmartens/asp-notes.pdf): the relaxed choice rule allows no assignment; the error rule forces a choice where possible (but, in keeping with the other constraints, allows generating an error fact when that is not possible).

scheibelp committed:
some experimental concretize.lp constraints (which don't work)

scheibelp committed:
experiment with replacing count-based constraint on version (to see if I'm doing something obviously-wrong with condition groups) - it doesn't seem to be an issue

scheibelp committed:
Revert "experiment with replacing count-based constraint on version (to see if I'm doing something obviously-wrong with condition groups) - it doesn't seem to be an issue"

This reverts commit 337484c955981faab851d1c8ed55065a82c9db85.

scheibelp committed:
Don't use arbitrary number for condition ID generation for condition groups

scheibelp committed:
Retain choice rule but don't apply cardinality constraint

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -472,6 +472,31 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
 #defined external_spec_condition/4.
 #defined external_spec_condition/5.
 
+%-----------------------------------------------------------------------------
+% Config required semantics
+%-----------------------------------------------------------------------------
```
Both `condition_group` and `condition_group_member` need to be `#defined` to avoid warnings like:
```
/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/concretize.lp:505:24-52: info: atom does not occur in any rule head:
  condition_group_member(Y,X)

/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/concretize.lp:505:73-100: info: atom does not occur in any rule head:
  condition_group(Package,X)

/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/concretize.lp:509:3-30: info: atom does not occur in any rule head:
  condition_group(Package,X)

/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/concretize.lp:510:3-31: info: atom does not occur in any rule head:
  condition_group_member(Y,X)

/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/concretize.lp:515:3-30: info: atom does not occur in any rule head:
  condition_group(Package,X)

/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/concretize.lp:521:53-81: info: atom does not occur in any rule head:
  condition_group_member(Y,X)

/home/culpo/PycharmProjects/spack/lib/spack/spack/solver/concretize.lp:521:98-125: info: atom does not occur in any rule head:
  condition_group(Package,X)
```

scheibelp replied:
This is fixed by https://github.com/spack/spack/pull/27987/commits/4ae6dcd5221fea7a0081e5e1a9b614f9479dead6

alalazo commented:
My understanding for the semantic is that (please correct me if I am wrong):
1. Within a single `one_of` list of constraints, only one of them must hold
2. Multiple items under `require` enforce each a single constraint from their list

For instance, with the following configuration:
```yaml
packages:
  zlib:
    require:
    - one_of: ['@1.2.12']
    - one_of: ['%gcc']
  libpng:
    require:
    - one_of: ['%gcc']
```
I expect all the constraints to be enforced, if the nodes are present. However, this is not the case if I install `zlib@1.2.11`:
```console
$ spack find
==> 1 installed package
-- linux-ubuntu20.04-icelake / gcc@9.4.0 ------------------------
zlib@1.2.11

$ spack spec libpng
Input spec
--------------------------------
libpng

Concretized
--------------------------------
libpng@1.6.37%gcc@9.4.0 arch=linux-ubuntu20.04-icelake
    ^zlib@1.2.11%gcc@9.4.0+optimize+pic+shared arch=linux-ubuntu20.04-icelake
```
Am I missing something?

scheibelp committed:
add defines to avoid clingo warnings

scheibelp committed:
fix issue when specifying 'require' as a string rather than a list

scheibelp committed:
additional example for zlib

scheibelp committed:
Condition groups need to include package-name in identity because they are generated on a per-package basis; also make weights independent of condition ids (this part isn't essential)

scheibelp committed:
adding initial test and module for config requirements

scheibelp committed:
update style for test

scheibelp committed:
fix line length

scheibelp committed:
another simple test

scheibelp committed:
another simple test

scheibelp committed:
additional test for selecting 'one_of'

scheibelp committed:
style edits

scheibelp committed:
sorting imports

scheibelp committed:
skipp config_requirements unit tests on Windows (as it uses the old concretizer)

scheibelp committed:
now that there are tests with examples, I'm removing the extra directory of examples

scheibelp committed:
need to import sys to check if we're on windows

scheibelp committed:
No functional/logic change: this is a rename of 'condition_weight' to 'requirement_weight' to avoid overloading the term 'condition' as it is used in concretize.lp

scheibelp committed:
reorder imports for style checks

scheibelp committed:
str-ify tmpdir paths for Python 2.7

scheibelp committed:
add skip statements when testing with original concretizer

scheibelp committed:
(no logic changes) update comments and error message

scheibelp committed:
add comments to tests; rename some tests; add a test with multiple one_of groups for one package

scheibelp commented:
@alalazo I have added a significant number of tests including one to cover https://github.com/spack/spack/pull/27987#issuecomment-1189701122. I think I should brainstorm it more for another day or two, but I'm curious if what you think of the testing module overall.

Also, when speaking with @becker33 he pointed out that the term `condition_group` was possibly confusing in the context of `concretize.lp`, so I renamed it to `requirement_group`.

scheibelp committed:
add another mock package and another test

scheibelp committed:
tweak comment

scheibelp committed:
add test to check prioritization of 'one_of' elements

scheibelp committed:
minor renaming

scheibelp commented:
Close/reopen to retry doc checks

scheibelp committed:
Merge branch 'develop' into features/config-required-versions

scheibelp committed:
relocate reused fixture to conftest

mwkrentel commented:
@scheibelp @tgamblin @alalazo @adamjstewart 
First, let me describe our normal use case.  We walk up to some
machine, make a fresh clone of spack develop, edit a `packages.yaml`
file and run `spack install hpctoolkit`, then usually copy the modules
file to some system directory.

Btw, we keep a "reference" `packages.yaml` file in the hpctoolkit
repo.  It's mostly the latest versions of packages with lots of
comments about what you might want to tweak on some systems.  It may
be overkill, but so far we use it.

In this context, we start with packages.yaml and a fresh install, so
of course, spack obeys all the preferences.  I think this is part of
the confusion.  If there are no already installed packages, then
there's nothing to compete with packages.yaml, so people expect that
spack will always obey it.

Then, 3 months go by, we release a new hpctoolkit, we git pull the
spack repo, update a few package versions and rerun `spack spec` and
don't get what we expect.  Hpctoolkit works closely with some packages
(dyninst, elfutils, libmonitor, intel-xed) and we want the very latest
versions of these, but for many other packages, most any version is Ok.

With `--reuse`, the reuse overrides packages.yaml.  With `--fresh`, it
rebuilds a lot more packages than necessary.  There isn't a definitive
way to tell spack exactly what to build, at least not outside of
environments.

I tried this PR and yes, it gives me a way to tell spack, "here is
what I want, don't override me."  I think something like this will be
necessary, so in general I approve.

mwkrentel commented:
There are some things about the syntax that are either not clear or
where I'm not sure it's the best choice.

1. I guess `require` takes a spec, whereas `version` and `variants`
accept just that.

2. I didn't figure out that `require` needs an `@` but version doesn't
until I first figured out (1) above.

3. In `require`, the argument needs to be quoted, but version does not.

4. I guess `require` takes a singleton, but version takes a list.
I guess that's what `one_of` is for.  But would it make more sense for
`require` to just take a list?

But the important thing is that I can add specs that spack must obey
and do so on a package-by-package basis.

mwkrentel commented:
Here's a wild and crazy idea.

I do suggest having preferences, but I didn't really explain our use
case very well.  Suppose the preferences were in the
`repos/builtin/packages/hpctoolkit` directory along side `package.py`.
That is we, as maintainers of hpctoolkit suggest building binutils
with `~nls` because we don't use it.  But any installer of hpctoolkit
is free to change it.

Essentially, we're saying, if you build hpctoolkit, we'd like to
change the defaults in the following dependency packages, but they're
just defaults, any installer is free to override the choices.  And
yes, the primary use case for this is to avoid building things that we
don't need.  But someone else is free to add it, if they want.

The idea is: make `packages.yaml` ALWAYS override reuse/fresh and have a
separate `preferences.yaml` file that belongs to the package.

An individual user doesn't really need to set a preference, he'd
rather just make the decision and set the value explicitly.  It makes
more sense for the package to set a preference.

----------

Btw, on an unrelated note, it would be nice to specify `arch` in a
`packages.yaml` file.  I can add `target`, but not `arch = platform-os-target`.

This happens on Cray where we need to build for the front end, so we
have to write `arch=cray-fe-x86_64` on the command line.

scheibelp committed:
remove unused import (since related fixture moved)

scheibelp commented:
> Btw, on an unrelated note, it would be nice to specify arch in a packages.yaml file. I can add target, but not arch = platform-os-target.

> This happens on Cray where we need to build for the front end, so we have to write arch=cray-fe-x86_64 on the command line.

@mwkrentel FWIW this should be possible with `require` e.g.:

```
packages:
  cmake:
    require: "arch=cray-fe-x86-64"
```

should work (if it doesn't, that's a bug with this PR). Some of the difference in the syntax is that the existing `packages.yaml` configuration (i.e. before this PR) has special-case configuration handling for different properties (e.g. as you pointed out, you can enter a "version" for a package in the config and this is understood at the level of the configuration itself). whereas with this PR, it uses generic spec syntax (anything you can pass to `Spec.satisfies()`, you can specify in `require:`).

mwkrentel commented:
@scheibelp I would need the `arch` spec under `all:`, like how we
can write `target` now:

```
packages:
   all:
      arch:  cray-fe-x86_64
```

scheibelp commented:
@alalazo FYI I think this is ready for review and sufficiently-tested

mwkrentel commented:
> I think this is ready for review and sufficiently-tested

I agree.

1. The important thing is that this provides a way to write a spec
that spack must obey.  This fills an important need.

2. It's a little odd to mix full specs for requires with partial specs
for version/variants.  But this is minor, (1) is the important thing.

I'm curious to see how this is used and evolves.  It may be everyone
writes all requires and gives up on the version/variants as mere
preferences.

alalazo commented:
> FYI I think this is ready for review and sufficiently-tested

@scheibelp Back from holiday, I'll review this PR tomorrow.

alalazo commented:
> Hpctoolkit works closely with some packages (dyninst, elfutils, libmonitor, intel-xed) and we want the very latest versions of these, but for many other packages, most any version is Ok.

Just a note that, if these are hard-requirements from the `hpctoolkit` package, they should be encoded in `package.py` and not in `packages.yaml`. Looking at the recipe, it seems that `dyninst` is the only dependency mentioned above with a version contraint that changes with increasing `hpctoolkit` versions.

alalazo commented:
@scheibelp I have a few high-level questions, related to how we want this new feature to behave. I plan to look at the code in details asap once the conflicts are fixed. 

## Should requirements be imposed on conflicting spec literals?

Currently using this `packages.yaml`:
```yaml
packages:
  zlib:
    require: '@1.2.11'
```
results in errors when trying to concretize to any other version:
```console
$ spack solve zlib@1.2.12
==> Error: Cannot satisfy requirement group for package 'zlib'
```
Is that the behavior we want? Personally, I think this is fine (especially considering that going otherwise might complicate a lot the logic in the PR) but I also see that it would be the first time in which a configuration in `packages.yaml` beats what is requested from the command line. In any case I think this should be clearly documented. Do you agree with that?

## The semantic is more "any_of" than "one_of"

With only GCC available:
```console
$ spack compiler list
==> Available compilers
-- gcc ubuntu20.04-x86_64 ---------------------------------------
gcc@9.4.0
```
I tried to generate a conflict for my spec by adding:
```yaml
packages:
  zlib:
    require:
    - one_of: ['os=ubuntu20.04', '%gcc']
```
but saw that the logic seems to be more that of `any_of`, i.e. if one _or more_ of the requirements in the list are satisfied the condition group is satisfied. The concretization was:
```console
$ spack solve zlib
==> Best of 2 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                            Installed  ToBuild
  1         number of input specs not concretized                        -        0
  2         number of packages to build (vs. reuse)                      -        0
  3         deprecated versions used                                     0        0
  4         requirement weight                                           0        0
  5         version weight                                               0        0
  6         number of non-default variants (roots)                       0        0
  7         preferred providers for roots                                0        0
  8         default values of variants not being used (roots)            0        0
  9         number of non-default variants (non-roots)                   0        0
  10        preferred providers (non-roots)                              0        0
  11        compiler mismatches                                          0        0
  12        OS mismatches                                                0        0
  13        non-preferred OS's                                           0        0
  14        version badness                                              0        0
  15        default values of variants not being used (non-roots)        0        0
  16        non-preferred compilers                                      0        0
  17        target mismatches                                            0        0
  18        non-preferred targets                                        0        0

zlib@1.2.12%gcc@9.4.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-icelake
```

## Are virtual specs taken into account?

I tried to generate another conflict with this configuration:
```yaml
packages:
  mpi:
    require:
    - one_of: ['os=ubuntu22.04']
  openmpi:
    require:
    - one_of: ['os=ubuntu20.04']
```
but `openmpi` concretized correctly to `ubuntu20.04`. In general, it seems virtual specs are not taken into account[^1]. I think it's fine to start with this limitation, and maintain it as long as nobody requires otherwise, but it's also something that should be documented. Maybe we can add a config audit to spot `require` sections under virtual dependencies and tell users that they have no effect.

[^1]: I also tried to require `@3` under `mpi` and I was able to concretize `openmpi@1.0`

scheibelp committed:
merge from develop (minor conflicts in schema/packages.py and test/concretize_preferences.py)

adamjstewart commented:
I am very behind on this discussion and haven't read through the code or all comments. From what I can tell, this is meant to address the common complaint that settings in `packages.yaml` are suggestions, not requirements. It does this by adding a new `requires` keyword where you can pass in a spec instead of separate `versions`, `variants`, etc. keywords.

Assuming that the above is correct, my only complaint is that "suggestions" and "requirements" have a very different interface. Should we get rid of `versions`, `variants`, etc. and replace them with `suggests`? That might make it more clear how they differ. Alternatively, should we simply convert the existing keys to be hard requirements instead of suggestions? Or bump them up in concretizer priority so that they are required unless it's physically impossible to achieve this (and overridden by CLI settings)?

Again, apologies if the above has already been asked/answered.

scheibelp commented:
@alalazo

> Should requirements be imposed on conflicting spec literals?

IMO yes

> The semantic is more "any_of" than "one_of"

That's true although I usually say "exactly one of" in the same sense that I say "XOR" or "exclusive-or" if I don't want the ambiguity of saying "or". The original suggestion for `one_of` is from https://github.com/spack/spack/pull/27987#discussion_r779783726 - I'm fine with renaming it to "any_of" though.

> [should] virtual specs [be] taken into account?

Yes although this essentially reuses our existing logic for applying constraints, so I don't think the problems here are unique to this PR. For example

* I have `%clang` and `%gcc` on my MacOS system
* I do `spack spec mpi` and it yields `openmpi%clang` (actually `%apple-clang`)
* If I do `spack spec mpi%gcc`, I still get `mpi%clang`

scheibelp commented:
> Assuming that the above is correct,

it is

> my only complaint is that "suggestions" and "requirements" have a very different interface. Should we get rid of versions, variants, etc. and replace them with suggests? 

I think we should, but in another PR. It was also suggested I could make an entirely separate config file for requirements (which might be simpler in the meantime and I think would be reasonable to do in this PR). I think it would make sense to have a transition period where the hold schema for package preferences is deprecated (similar to other parts of config that have changed recently).

adamjstewart commented:
Awesome! This will be super useful. I've always wanted `requires` instead of `suggests`. It's been frustrating to see Spack ignore my requirements in environments, especially now that we have `--reuse` by default.

alalazo commented:
> That's true although I usually say "exactly one of" in the same sense that I say "XOR" or "exclusive-or" if I don't want the ambiguity of saying "or".

Ok, I was confused because in https://github.com/spack/spack/pull/27987#issuecomment-1187487805 and following comment it seemed to me that `one_of` meant "exactly one of". I would be fine with either changing the semantic to "exactly one of" or changing the name to `any_of`. [^1] I think they are both powerful enough to express what the user wants.[^2]

[^1]: The semantic then would match e.g. Python's `any` or [JSONSCHEMA](https://json-schema.org/understanding-json-schema/reference/combining.html#anyof).
[^2]: I'll try to come up with counter-examples, but I doubt I'll find any.

>  The original suggestion for one_of is from https://github.com/spack/spack/pull/27987#discussion_r779783726 - I'm fine with renaming it to "any_of" though.

In https://github.com/spack/spack/pull/27987#discussion_r779783726 the example `yaml` was constructed to have items in each list which are mutually exclusive, in which case `one_of` behaves the same as `any_of`. To be clear, what I'm trying to do at this point is to clarify the semantic, so that we know how this mechanism should behave with an arbitrary list of specs in each item.

> If I do spack spec mpi%gcc, I still get mpi%clang

Fair, but if you do `mpi@3` you shouldn't be able to concretize to `openmpi@1.X` as I did above. Anyhow, I don't see the edge cases with virtuals as a stopper for this PR. I think we should just document that `require` is not meant to be used with them for the time being. We can extend the semantic of `require:` to virtuals in a follow up PR, and discuss use cases that we currently cannot cover. For instance, "require the provider for `mpi` to be `openmpi@X.Y +cuda`".

scheibelp committed:
run style fix

alalazo commented on lib/spack/spack/schema/packages.py:
```diff
@@ -21,6 +21,20 @@
+                                "type": "array",
+                                "items": {
+                                    "type": "object",
+                                    "properties": {"one_of": {"type": "array"}},
```
As mentioned in https://github.com/spack/spack/pull/27987#issuecomment-1204396111 I would either go with:
```suggestion
                                    "properties": {"any_of": {"type": "array"}},
```
or change the semantics to "exactly one of".

alalazo replied:
If the bugfix below works, it might be easier to change to "exactly one of".

alalazo replied:
Also, if we're using "exactly one of" we'll have less trouble with requirement weights (it will be one weight per group per member)

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -878,6 +878,31 @@ def package_compiler_defaults(self, pkg):
+            rules = []
+            for requirement in requirements:
+                rules.append((pkg_name, requirement["one_of"]))
+        cnd_grp_id = 0
```
(minor) Maybe this is slightly more readable:
```suggestion
        group_id = 0
```
?

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -878,6 +878,31 @@ def package_compiler_defaults(self, pkg):
+            for requirement in requirements:
+                rules.append((pkg_name, requirement["one_of"]))
+        cnd_grp_id = 0
+        for pkg_name, cnd_group in rules:
```
Another minor naming suggestion:
```suggestion
        for pkg_name, requirement_group in rules:
```

alalazo replied:
(minor) Same here to avoid manual increment of `cnd_group_id`:
```suggestion
        for cnd_group_id, (pkg_name, cnd_group) in enumerate(rules):
```

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -878,6 +878,31 @@ def package_compiler_defaults(self, pkg):
+            self.gen.fact(fn.requirement_group(pkg_name, cnd_grp_id))
+            for spec_str in cnd_group:
+                spec = spack.spec.Spec(spec_str)
+                if not spec.name:
+                    spec.name = pkg_name
```
This is interesting, as it means we can do things like:
```yaml
packages:
  hdf5:
    require:
    - one_of: ['trilinos@X.Y']
```
where `trilinos` can be an arbitrary package. It seems fine to me, wondering if there are "smart" ways to use this.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -878,6 +878,31 @@ def package_compiler_defaults(self, pkg):
                 fn.node_compiler_preference(pkg.name, cspec.name, cspec.version, -i * 100)
             )
 
+    def cfg_required_rules(self, pkg):
```
(minor) A name similar to:
```suggestion
    def package_requirement_rules(self, pkg):
```
seems more in line with other methods already present.

alalazo commented on lib/spack/spack/solver/asp.py:
```diff
@@ -878,6 +878,31 @@ def package_compiler_defaults(self, pkg):
+        for pkg_name, cnd_group in rules:
+            requirement_weight = 0
+            self.gen.fact(fn.requirement_group(pkg_name, cnd_grp_id))
+            for spec_str in cnd_group:
```
(minor) I think you can avoid manual increment and initialization of `requirement_weight` by:
```suggestion
            for requirement_weight, spec_str in enumerate(cnd_group):
```

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -495,6 +495,39 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
+% A package may have an associated "requirement_group": a set of specs where
+% at least one must be applied. Each spec in the group is associated
+% with a condition via asp.impose().
+{ condition_holds(Y) : requirement_group_member(Y, Package, X) } :-
+  node(Package),
+  requirement_group(Package, X).
```
(bug) Here I think we have a logical error that I need to think how to solve. The problem is that we might fulfill all the conditions to `impose(Y)` without clingo deciding to create a `condition_holds(Y)` atom. This is related to `condition_holds` appearing in a choice rather than a rule. To be more clear, if the requirement is `zlib@1.2.12`, clingo might decide not to have the corresponding `condition_holds` atom in the answer set, but still derive `version(zlib, 1.2.12)` by other means.

alalazo replied:
@scheibelp I have to run unit-tests, but I think this is a possible solution to the bug:
```diff
diff --git a/lib/spack/spack/solver/asp.py b/lib/spack/spack/solver/asp.py
index 31d765feaa..dde7898c3a 100644
--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -945,10 +945,9 @@ def cfg_required_rules(self, pkg):
                 spec = spack.spec.Spec(spec_str)
                 if not spec.name:
                     spec.name = pkg_name
-                member_id = next(self._condition_id_counter)
+                member_id = self.condition(spec, imposed_spec=spec, name=pkg_name)
                 self.gen.fact(fn.requirement_group_member(member_id, pkg_name, cnd_grp_id))
                 self.gen.fact(fn.requirement_has_weight(member_id, requirement_weight))
-                self.impose(member_id, spec, name=pkg_name, node=False)
                 requirement_weight += 1
             cnd_grp_id += 1
 
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 17ace1aa5a..deca651868 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -504,18 +504,10 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
 % Config required semantics
 %-----------------------------------------------------------------------------
 
-% A package may have an associated "requirement_group": a set of specs where
-% at least one must be applied. Each spec in the group is associated
-% with a condition via asp.impose().
-{ condition_holds(Y) : requirement_group_member(Y, Package, X) } :-
-  node(Package),
-  requirement_group(Package, X).
-
 requirement_group_satisfied(Package, X) :-
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) } ,
   node(Package),
-  requirement_group(Package, X),
-  requirement_group_member(Y, Package, X),
-  condition_holds(Y).
+  requirement_group(Package, X).
 
 error(2, "Cannot satisfy requirement group for package '{0}'", Package) :-
   node(Package),

```

alalazo replied:
Not that above you can change the semantics from `any_of` to `one_of` by just adding an upper bound to the cardinality constraint:
```
 1 { condition_holds(Y) : requirement_group_member(Y, Package, X) } 1
```

alalazo replied:
I have run tests succesfully on:
* **Spack:** 0.19.0.dev0 (8577f4c9401f14d6d017eeb936b1b62bc246c63b)
* **Python:** 3.8.10
* **Platform:** linux-ubuntu20.04-icelake
* **Concretizer:** clingo

scheibelp replied:
> clingo might decide not to have the corresponding condition_holds atom in the answer set, but still derive version(zlib, 1.2.12) by other means.

That is true, but I considered that fine: if you combine my original constraint with the below constraints that generate error facts, my intent was that an error would be generated if you cannot derive `condition_holds(Y)`.

I originally used cardinality constraints in this rule, but shifted to this approach in light of https://github.com/spack/spack/pull/30669/files#diff-09156cc9ff0b92ff5a0082e65d0c6ca11ecb86b1af9e719ff29d8445a0a8d332.

If we applied your changes, I assume the below rules like 

```
error(2, "Cannot satisfy requirement group for package '{0}'", Package) :-
   node(Package),
   requirement_group(Package, X),
   not requirement_group_satisfied(Package, X).
```

would never be triggered (i.e. the concretizer would just say it couldn't find a solution vs. producing a solution with error facts) - is that true? Or does your approach also make use of the error facts?

> Not that above you can change the semantics from any_of to one_of by just adding an upper bound to the cardinality constraint:

To be clear I think you mean that when you see `one_of` you interpret it to mean "exactly one of" (I didn't interpret it that way, but I can proceed with that interpretation). I don't think "exactly-one-of" semantics are desirable: I think this might drive the concretizer away from sensible solutions that happen to satisfy more than one spec in a requirement group (and I don't think users would ever depend on the mutual exclusion).

Therefore I should rename this to "any_of" (which I originally avoided because in other contexts - e.g. variants - "any of" also allows for choosing none of the options) or perhaps even better: "at_least_one_of" (which I think would eliminate all ambiguities).

alalazo replied:
> is that true? Or does your approach also make use of the error facts?

The approach is only a slight variation of yours and generates error facts. With a "exactly one of" semantic and this:
```yaml
packages:
  zlib:
    require:
    - one_of: ['%gcc', '%gcc']
```
I get:
```console
$ spack spec zlib
==> Error: Cannot satisfy requirement group for package 'zlib'
```
The point is that I use a cardinality constraint in the body of the rule to derive `requirement_group_satisfied(Package, X)`, and I enforce the full condition.

I think the current approach is inconsistent internally. You will not have `condition_holds(Y)` for conditions that actually hold, and the weight being used is that of the first constraint matching. So at least we have a mismatch between the name of the atom and its meaning.

Using `one_of` (in the sense of "exactly one of") can lead to a more straightforward way of dealing with weights, and I think gives the same expressiveness to users.

alalazo commented on lib/spack/spack/solver/concretize.lp:
```diff
@@ -1088,156 +1121,168 @@ opt_criterion(100, "number of packages to build (vs. reuse)").
-opt_criterion(15, "deprecated versions used").
-#minimize{ 0@215: #true }.
-#minimize{ 0@15: #true }.
+opt_criterion(75, "deprecated versions used").
```
Can you explain why all this renumbering was necessary?

scheibelp replied:
At one point, I had prioritized this differently and found myself in a scenario akin to `ArrayList.insert(0, item)` except I was the computer. I created "space" between these priority rankings so that other users can insert new constraints wherever they want  without having to shift everything by 1 (at some point this isn't possible, but I figured I'd help out the next 2-3 people).

alalazo commented on lib/spack/spack/test/config_requirements.py:
```diff
@@ -0,0 +1,235 @@
+# Copyright 2013-2022 Lawrence Livermore National Security, LLC and other
```
(minor) Should we use `concretize_requirements.py` (as opposed to `concretize_preferences.py`)?

alalazo commented on lib/spack/spack/test/config_requirements.py:
```diff
@@ -0,0 +1,235 @@
+    spack.config.set("packages", conf["packages"], scope="concretize")
+
+
+_pkgx = (
+    "x",
+    """\
+class X(Package):
+    version('1.1')
+    version('1.0')
+    version('0.9')
+
+    variant('shared', default=True,
+            description='Build shared libraries')
+
+    depends_on('y')
+""",
+)
```
(comment) I wonder if we can generalize this, and use Jinja2 and templates. Anyhow, this is for sure something for after this PR.

alalazo reviewed (changes requested):
I think I found a bug in the logic program. I'll try to think how to fix it, and get back here if I have an idea. Besides that only minor points.

alalazo commented:
Here's a diff with both the `any_of` and the `one_of` semantics:
```diff
diff --git a/lib/spack/spack/schema/packages.py b/lib/spack/spack/schema/packages.py
index 9b25b5b495..fa5842b38e 100644
--- a/lib/spack/spack/schema/packages.py
+++ b/lib/spack/spack/schema/packages.py
@@ -28,7 +28,10 @@
                                 "type": "array",
                                 "items": {
                                     "type": "object",
-                                    "properties": {"one_of": {"type": "array"}},
+                                    "properties": {
+                                        "one_of": {"type": "array"},
+                                        "any_of": {"type": "array"},
+                                    },
                                 },
                             },
                             # Shorthand for a condition group with one member
diff --git a/lib/spack/spack/solver/asp.py b/lib/spack/spack/solver/asp.py
index 31d765feaa..d882249f44 100644
--- a/lib/spack/spack/solver/asp.py
+++ b/lib/spack/spack/solver/asp.py
@@ -932,25 +932,24 @@ def cfg_required_rules(self, pkg):
         config = spack.config.get("packages")
         requirements = config.get(pkg_name, {}).get("require", {})
         if isinstance(requirements, string_types):
-            rules = [(pkg_name, [requirements])]
+            rules = [(pkg_name, "one_of", [requirements])]
         else:
             rules = []
             for requirement in requirements:
-                rules.append((pkg_name, requirement["one_of"]))
-        cnd_grp_id = 0
-        for pkg_name, cnd_group in rules:
-            requirement_weight = 0
+                for policy in ("one_of", "any_of"):
+                    if policy in requirement:
+                        rules.append((pkg_name, policy, requirement[policy]))
+
+        for cnd_grp_id, (pkg_name, policy, cnd_group) in enumerate(rules):
             self.gen.fact(fn.requirement_group(pkg_name, cnd_grp_id))
-            for spec_str in cnd_group:
+            self.gen.fact(fn.requirement_policy(pkg_name, cnd_grp_id, policy))
+            for requirement_weight, spec_str in enumerate(cnd_group):
                 spec = spack.spec.Spec(spec_str)
                 if not spec.name:
                     spec.name = pkg_name
-                member_id = next(self._condition_id_counter)
+                member_id = self.condition(spec, imposed_spec=spec, name=pkg_name)
                 self.gen.fact(fn.requirement_group_member(member_id, pkg_name, cnd_grp_id))
                 self.gen.fact(fn.requirement_has_weight(member_id, requirement_weight))
-                self.impose(member_id, spec, name=pkg_name, node=False)
-                requirement_weight += 1
-            cnd_grp_id += 1
 
     def pkg_rules(self, pkg, tests):
         pkg = packagize(pkg)
diff --git a/lib/spack/spack/solver/concretize.lp b/lib/spack/spack/solver/concretize.lp
index 17ace1aa5a..2438898963 100644
--- a/lib/spack/spack/solver/concretize.lp
+++ b/lib/spack/spack/solver/concretize.lp
@@ -504,31 +504,41 @@ error(2, "Attempted to use external for '{0}' which does not satisfy any configu
 % Config required semantics
 %-----------------------------------------------------------------------------
 
-% A package may have an associated "requirement_group": a set of specs where
-% at least one must be applied. Each spec in the group is associated
-% with a condition via asp.impose().
-{ condition_holds(Y) : requirement_group_member(Y, Package, X) } :-
+requirement_group_satisfied(Package, X) :-
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) } 1,
   node(Package),
+  requirement_policy(Package, X, "one_of"),
   requirement_group(Package, X).
 
+requirement_weight(Package, W) :-
+  condition_holds(Y),
+  requirement_has_weight(Y, W),
+  requirement_group_member(Y, Package, X),
+  requirement_policy(Package, X, "one_of"),
+  requirement_group_satisfied(Package, X).
+
 requirement_group_satisfied(Package, X) :-
+  1 { condition_holds(Y) : requirement_group_member(Y, Package, X) } ,
   node(Package),
-  requirement_group(Package, X),
-  requirement_group_member(Y, Package, X),
-  condition_holds(Y).
+  requirement_policy(Package, X, "any_of"),
+  requirement_group(Package, X).
+
+requirement_weight(Package, W) :-
+  W = #min {
+    Z : requirement_has_weight(Y, Z), condition_holds(Y), requirement_group_member(Y, Package, X);
+    % We need this to avoid an annoying warning during the solve
+    %   concretize.lp:1151:5-11: info: tuple ignored:
+    %   #sup@73
+    10000
+  },
+  requirement_policy(Package, X, "any_of"),
+  requirement_group_satisfied(Package, X).
 
 error(2, "Cannot satisfy requirement group for package '{0}'", Package) :-
   node(Package),
   requirement_group(Package, X),
   not requirement_group_satisfied(Package, X).
 
-requirement_weight(Package, W) :-
-  condition_holds(Y),
-  requirement_has_weight(Y, W),
-  requirement_group_member(Y, Package, X),
-  node(Package),
-  requirement_group(Package, X).
-
 #defined requirement_group/2.
 #defined requirement_group_member/3.
 #defined requirement_has_weight/2.

```

alalazo commented:
With the following `packages.yaml`:
```yaml
packages:
  zlib:
    require:
    - any_of: ['@1.2.11', '%gcc']   
```
Currently Spack will concretize to version `1.2.12` because requirements from `packages.yaml` are lower priority than avoiding deprecated versions (and zlib@1.2.11` is deprecated). I think we might want to switch those two optimizations?

alalazo commented:
Another buggy use case reproducer. Start with:
```console
$ spack compiler list
==> Available compilers
-- gcc ubuntu20.04-x86_64 ---------------------------------------
gcc@9.4.0

$ spack install zlib
==> Installing zlib-1.2.12-6jwud6k4455g42yxfqsmjnco2olbupjg
==> No binary for zlib-1.2.12-6jwud6k4455g42yxfqsmjnco2olbupjg found: installing from source
==> Using cached archive: /home/culpo/PycharmProjects/spack/var/spack/cache/_source-cache/archive/91/91844808532e5ce316b3c010929493c0244f3d37593afd6de04f71821d5136d9.tar.gz
==> Applied patch /home/culpo/PycharmProjects/spack/var/spack/repos/builtin/packages/zlib/configure-cc.patch
==> zlib: Executing phase: 'install'
==> zlib: Successfully installed zlib-1.2.12-6jwud6k4455g42yxfqsmjnco2olbupjg
  Fetch: 0.01s.  Build: 1.72s.  Total: 1.73s.
[+] /home/culpo/PycharmProjects/spack/opt/spack/linux-ubuntu20.04-icelake/gcc-9.4.0/zlib-1.2.12-6jwud6k4455g42yxfqsmjnco2olbupjg
```
Then add a `packages.yaml` like:
```yaml
packages:
  zlib:
    require:
    - one_of: ['@1.2.12', '%gcc']
```
If we concretize:
```
$ spack spec zlib
Input spec
--------------------------------
zlib

Concretized
--------------------------------
zlib@1.2.12%gcc@9.4.0+optimize+pic+shared patches=0d38234 arch=linux-ubuntu20.04-icelake
```
we get a spec satisfying both conditions. This happens because the spec is reused, and I think for reused spec we are not able to derive that the `%gcc` condition holds. Using instead:
```console
$ spack spec --fresh zlib
==> Warning: using "zlib@1.2.11" which is a deprecated version
Input spec
--------------------------------
zlib

Concretized
--------------------------------
zlib@1.2.11%gcc@9.4.0+optimize+pic+shared arch=linux-ubuntu20.04-icelake
```
Gives the expected result.

scheibelp commented:
Regarding https://github.com/spack/spack/pull/27987#issuecomment-1210983266:

(Note: "one_of" as it is treated in this point of the PR will be renamed to "any_of", and I assume we are talking about the semantics of "any_of")

> Another buggy use case reproducer. Start with:

I think that like https://github.com/spack/spack/pull/27987#issuecomment-1210962058, this is just a matter of the current prioritization: reuse has higher priority than condition requirements.

I don't actually consider this particularly confusing either: if the user has specified

```
packages:
  zlib:
    require:
    - one_of: ['@1.2.12', '%gcc']
```

They are saying "I need zlib to be 1.2.12, but I can also use other zlibs if %gcc is used". If they would be disappointed with any zlib but 1.2.12, they can create a single entry that just specifies 1.2.12. In other words, my intent (and the tests) primarily are intended to confirm that we have a sensible interpretation of the prioritization of elements within a `one_of` in the absence of other prioritization influences.

That all being said. I'm fine with rearranging the priorities.

alalazo commented:
@scheibelp 

> (Note: "one_of" as it is treated in this point of the PR will be renamed to "any_of", and I assume we are talking about the semantics of "any_of")

Apologies for the confusion. I tested this on top of my PR, so `one_of` means "exactly one of". I debugged with `clingo` and it seems we miss some intermediate atom (iirc `node_compiler_version_satisfies`) when we reuse a spec vs. when we build it. I still need to dig deeper though.

scheibelp commented:
Could https://github.com/spack/spack/pull/27987#issuecomment-1210983266 be because `_facts_from_concrete_spec` calls `impose` rather than `condition`? Our discussion in https://github.com/spack/spack/pull/27987#discussion_r939001357 gives me the impression that impose should never be called by itself (essentially we want "if-and-only-if", and `impose` only enforces the "if" part).

scheibelp committed:
changes from PR

scheibelp committed:
create test that should fail but passes

scheibelp committed:
with some adjustments, the test now fails as expected

scheibelp committed:
remove commented code

scheibelp committed:
new added test is fixed by condition(), which transforms 'if' to 'iff' constraint

scheibelp committed:
auto style fix

scheibelp committed:
str-ify for python 2.7

scheibelp commented:
@alalazo I added https://github.com/scheibelp/spack/pull/21 here, added a test which reproduces https://github.com/spack/spack/pull/27987#issuecomment-1210983266, and fixed it in https://github.com/spack/spack/pull/27987/commits/1d733afd795ac65654141fa17729c9454544a813

That commit however breaks reuse tests that check that you can reuse a package installation when the underlying package.py is changed (e.g. `test_reuse_installed_packages_when_package_def_changes`), so I'd need to investigate this more tomorrow/Monday.

scheibelp commented:
I don't think https://github.com/spack/spack/commit/1d733afd795ac65654141fa17729c9454544a813 makes sense (or at least, it doesn't seem like it should be necessary to enforce the constraints we want). The test should still be useful (I'll be using it to debug the issue with "one_of").

alalazo commented:
> I don't think https://github.com/spack/spack/commit/1d733afd795ac65654141fa17729c9454544a813 makes sense (or at least, it doesn't seem like it should be necessary to enforce the constraints we want).

I agree. Concrete specs derive the `impose(Hash)` from the `hash(Package, Hash)` atom, which is a choice of clingo to pop into existence if possible. Abstract specs instead derive `impose(ID)` from the fact that `condition_holds(ID)`, without any choice involved. I think we just need to reconstruct some intermediate atoms which are not in the answer set for concrete specs.

alalazo committed:
Fix failing unit-tests by deriving an atom for concretized specs

Also switch requirements and deprecated versions priority

alalazo commented:
These modifications should fix the unit-tests and take care of https://github.com/spack/spack/pull/27987#issuecomment-1210962058 : https://github.com/scheibelp/spack/pull/23

scheibelp committed:
adjust naming according to review

scheibelp committed:
rename test module

scheibelp committed:
add documentation

scheibelp committed:
add clarification to doc

scheibelp committed:
update schema comments now that property names have chnaged

alalazo commented on lib/spack/docs/build_settings.rst:
```diff
@@ -339,6 +339,58 @@ concretization rules.  A provider lists a value that packages may
+
+* You can only specify requirements for specific packages: you cannot
+  add ``requires`` under ``all``.
+* The order of specs in ``any_of`` indicates a preference: items that
+  appear earlier in the list are preferred (note that these
+  preferences can be ignored in favor of others).
```
This is true also for `one_of`. Being a very minor update, I would submit a separate PR once this one is merged, if you agree with that.

alalazo commented on lib/spack/docs/build_settings.rst:
```diff
@@ -339,6 +339,58 @@ concretization rules.  A provider lists a value that packages may
+
+Other notes about ``requires``:
+
+* You can only specify requirements for specific packages: you cannot
+  add ``requires`` under ``all``.
```
At the moment also `require` under virtual specs are disregarded. As above, I would clarify in a follow-up PR.

haampie commented on lib/spack/docs/build_settings.rst:
```diff
@@ -339,6 +339,70 @@ concretization rules.  A provider lists a value that packages may
+       - any_of: ["~cuda", "gcc"]
+     mpich:
+      require:
+      - one of: ["+cuda", "+rocm"]
```
```suggestion
      - one_of: ["+cuda", "+rocm"]
```

scheibelp committed:
doc updates

tgamblin reviewed (approved):
@scheibelp this looks great -- thanks!

haampie commented on lib/spack/docs/build_settings.rst:
```diff
@@ -339,6 +339,72 @@ concretization rules.  A provider lists a value that packages may
+       require: "@1.13.2"
+     openmpi:
+       require:
+       - any_of: ["~cuda", "gcc"]
```
```suggestion
       - any_of: ["~cuda", "%gcc"]
```

haampie commented:
I'm somewhat afraid one_of/any_of can be somewhat confusing. Maybe what's missing is some docs on how to combine one_of/any_of with unconditional requirements, like always use `@4:`.

scheibelp commented:
Close/reopen to restart docs build

scheibelp committed:
Merge branch 'develop' into features/config-required-versions

tgamblin commented:
We should do a follow-on doc PR to address @haampie's concerns, but this LGTM.
